pkg/rehearse/jobs.go

```diff
@@ -168,89 +169,99 @@ func inlineCiOpConfig(job *prowconfig.Presubmit, targetRepo string, ciopConfigs
 
 	for repo, jobs := range toBeRehearsed {
 		for _, job := range jobs {
-			fields := logrus.Fields{"target-repo": repo, "target-job": job.Name}
-			jobLogger := Loggers{loggers.Job.WithFields(fields), loggers.Debug.WithFields(fields)}
+			jobLogger := loggers.Job.WithFields(logrus.Fields{"target-repo": repo, "target-job": job.Name})
 			rehearsal, err := makeRehearsalPresubmit(&job, repo, prNumber)
 			if err != nil {
-				jobLogger.Job.WithError(err).Warn("Failed to make a rehearsal presubmit")
+				jobLogger.WithError(err).Warn("Failed to make a rehearsal presubmit")
 				continue
 			}
 
-			rehearsal, err = inlineCiOpConfig(rehearsal, repo, ciopConfigs, jobLogger)
+			rehearsal, err = inlineCiOpConfig(rehearsal, repo, ciopConfigs, loggers)
 			if err != nil {
-				jobLogger.Job.WithError(err).Warn("Failed to inline ci-operator-config into rehearsal job")
+				jobLogger.WithError(err).Warn("Failed to inline ci-operator-config into rehearsal job")
 				continue
 			}
 
-			jobLogger.Job.WithField(LogRehearsalJob, rehearsal.Name).Info("Created a rehearsal job to be submitted")
+			jobLogger.WithField(LogRehearsalJob, rehearsal.Name).Info("Created a rehearsal job to be submitted")
 			rehearsals = append(rehearsals, rehearsal)
 		}
 	}
 
+	return rehearsals
+}
+
+type Executor struct {
+	dryRun     bool
+	rehearsals []*prowconfig.Presubmit
+	prNumber   int
+	prRepo     string
+	refs       *pjapi.Refs
+	loggers    Loggers
+	pjclient   pj.ProwJobInterface
+}
+
+func NewExecutor(toBeRehearsed map[string][]prowconfig.Presubmit, prNumber int, prRepo string, refs *pjapi.Refs,
+	dryRun bool, loggers Loggers, pjclient pj.ProwJobInterface) *Executor {
+	rehearsals := configureRehearsalJobs(toBeRehearsed, prRepo, prNumber, loggers)
+	return &Executor{
+		dryRun:     dryRun,
+		rehearsals: rehearsals,
+		prNumber:   prNumber,
+		prRepo:     prRepo,
+		refs:       refs,
+		loggers:    loggers,
+		pjclient:   pjclient,
+	}
+}
+
+// ExecuteJobs takes configs for a set of jobs which should be "rehearsed", and
+// creates the ProwJobs that perform the actual rehearsal. *Rehearsal* means
+// a "trial" execution of a Prow job configuration when the *job config* config
+// is changed, giving feedback to Prow config authors on how the changes of the
+// config would affect the "production" Prow jobs run on the actual target repos
+func (e *Executor) ExecuteJobs() (bool, error) {
 	submitSuccess := true
-	pjs := make(sets.String, len(rehearsals))
-	for _, job := range rehearsals {
-		created, err := submitRehearsal(job, refs, loggers, pjclient)
-		if err != nil {
-			loggers.Job.WithError(err).Warn("Failed to execute a rehearsal presubmit")
-			submitSuccess = false
-			continue
-		}
-		loggers.Job.WithFields(pjutil.ProwJobFields(created)).Info("Submitted rehearsal prowjob")
-		pjs.Insert(created.Name)
+	follow := !e.dryRun
+
+	pjs, err := e.submitRehearsals()
+	if err != nil {
+		submitSuccess = false
 	}
-	if !follow {
+
+	if follow {
```

Comment:
When `follow` is true, we *want* it to continue to the end of the function (to "follow" the jobs).

```suggestion
	if !follow {
```